test/TestUtilities/Test.Utility/SelfSignedCertificateMockServer.cs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NuGet.Test.Server;
+
+namespace Test.Utility
+{
+    public class SelfSignedCertificateMockServer
+    {
+        private readonly X509Certificate2 _certificate;
+        private readonly string _packageDirectory;
+        private TcpListener _tcpListener;
+        private bool _runServer = true;
+        public string URI;
+
+        public SelfSignedCertificateMockServer(string packageDirectory)
+        {
+            _packageDirectory = packageDirectory;
+            _certificate = GenerateSelfSignedCertificate();
+        }
+
+        public async Task StartServer()
+        {
+
+            var portReserver = new PortReserver();
+            var portNumber = await portReserver.ExecuteAsync((p, t) => Task.FromResult(p), CancellationToken.None);
+            _tcpListener = new TcpListener(IPAddress.Loopback, portNumber);
+            URI = $"https://{_tcpListener.LocalEndpoint}/";
+            _tcpListener.Start();
+
+            while (_runServer)
+            {
+                var client = await _tcpListener.AcceptTcpClientAsync();
+                _ = Task.Run(() => HandleClient(client));
+            }
+        }
+
+        public void StopServer()
+        {
+            _runServer = false;
+            _tcpListener.Stop();
+        }
+
+        private async Task HandleClient(TcpClient client)
+        {
+            using (client)
+            using (var sslStream = new SslStream(client.GetStream(), false))
+            {
+                await sslStream.AuthenticateAsServerAsync(_certificate, clientCertificateRequired: false, enabledSslProtocols: SslProtocols.Tls12, checkCertificateRevocation: true);
```

Comment:
I removed `enabledSslProtocols: SslProtocols.Tls12` argument because there is another overload that doesn't need this parameter and all the 4 new tests passed locally. I guess we can remove passing this parameter. The reason is, if Tls12 protocol gets deprecated in the future, we have to make a code change to use another protocol.